ssl/quic/quic_rx_depack.c

```diff
@@ -163,127 +33,178 @@ static void ossl_quic_fatal(QUIC_CONNECTION *c, int al, int reason,
         /* In case sizeof(uint64_t) > sizeof(size_t) */
         || total_ranges > SIZE_MAX / sizeof(ack_ranges[0])
         || (ack_ranges = OPENSSL_zalloc(sizeof(ack_ranges[0])
                                         * (size_t)total_ranges)) == NULL)
-        return 0;
+        goto malformed;
 
     ack.ack_ranges = ack_ranges;
     ack.num_ack_ranges = (size_t)total_ranges;
 
     if (!ossl_quic_wire_decode_frame_ack(pkt, ack_delay_exp, &ack, NULL))
-        ok = 0;
-    if (ok
-        && !ossl_ackm_on_rx_ack_frame(GET_CONN_ACKM(connection), &ack,
-                                      packet_space, received))
-        ok = 0;
+        goto malformed;
+
+    if (!ossl_ackm_on_rx_ack_frame(ch->ackm, &ack,
+                                   packet_space, received))
+        goto malformed;
 
     OPENSSL_free(ack_ranges);
-    if (!ok)
-        return 0;
     return 1;
+
+malformed:
+    ossl_quic_channel_raise_protocol_error(ch,
+                                           QUIC_ERR_FRAME_ENCODING_ERROR,
+                                           frame_type,
+                                           "decode error");
+    OPENSSL_free(ack_ranges);
+    return 0;
 }
 
 static int depack_do_frame_reset_stream(PACKET *pkt,
-                                        QUIC_CONNECTION *connection,
+                                        QUIC_CHANNEL *ch,
                                         OSSL_ACKM_RX_PKT *ackm_data)
 {
     OSSL_QUIC_FRAME_RESET_STREAM frame_data;
     QUIC_STREAM *stream = NULL;
-    int stream_type = 0;
 
-    if (!ossl_quic_wire_decode_frame_reset_stream(pkt, &frame_data))
+    if (!ossl_quic_wire_decode_frame_reset_stream(pkt, &frame_data)) {
+        ossl_quic_channel_raise_protocol_error(ch,
+                                               QUIC_ERR_FRAME_ENCODING_ERROR,
+                                               OSSL_QUIC_FRAME_TYPE_RESET_STREAM,
+                                               "decode error");
         return 0;
+    }
 
     /* This frame makes the packet ACK eliciting */
     ackm_data->is_ack_eliciting = 1;
 
-    if ((stream = ssl_get_stream(connection, frame_data.stream_id)) == NULL)
-        return 0;
-    stream_type = ssl_get_stream_type(stream);
-
-    ssl_close_stream(stream, SHUT_WR); /* Reuse shutdown(2) symbols */
-    if ((stream_type & SSL_STREAM_TYPE_S) != 0) {
-        QUICfatal(connection, QUIC_STREAM_STATE_ERROR, ERR_R_INTERNAL_ERROR);
+    stream = ossl_quic_stream_map_get_by_id(&ch->qsm, frame_data.stream_id);
+    if (stream == NULL || stream->rstream == NULL) {
+        ossl_quic_channel_raise_protocol_error(ch,
+                                               QUIC_ERR_STREAM_STATE_ERROR,
+                                               OSSL_QUIC_FRAME_TYPE_RESET_STREAM,
+                                               "RESET_STREAM frame for nonexistent or "
+                                               "TX only stream");
```

Comment:
Could we distinguish in the message whether the stream is nonexistent or TX only?